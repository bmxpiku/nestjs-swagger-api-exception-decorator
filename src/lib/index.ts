import { HttpException } from '@nestjs/common';

import { ApiException } from './decorators/api-exception.decorator';
import {
  ExceptionArguments,
  ExceptionOrExceptionArray,
  ExceptionOrExceptionArrayFunc,
} from './interfaces/api-exception.interface';
import { Options, Template } from './interfaces/options.interface';
import { areExceptionsPassedWithoutArrowFunction } from './utils/exception.util';

/**
 * Build your own custom decorator. This enables you to re-use the same template again without the need to specify it
 * over and over again
 *
 * @param template Any object describing the template which should be shown as example value
 * @param globalOptions Specify the content type
 */
export const buildTemplatedApiExceptionDecorator = <T = Template>(
  template: T,
  globalOptions?: Omit<Options<T>, 'template'>,
) => {
  return <Exception extends HttpException>(exceptions: ExceptionArguments<Exception>, options?: Options<T>) => {
    const mergedOptions = ({ ...globalOptions, template, ...options } as unknown) as Options<Template>;

    if (areExceptionsPassedWithoutArrowFunction(exceptions)) {
      return ApiException(exceptions as ExceptionOrExceptionArray<Exception>, mergedOptions);
    } else {
      return ApiException(exceptions as ExceptionOrExceptionArrayFunc<Exception>, mergedOptions);
    }
  };
};

export { ApiException };
