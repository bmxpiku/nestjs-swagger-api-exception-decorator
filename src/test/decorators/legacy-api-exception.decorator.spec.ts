// TODO: Remove this file if the deprecated decorator signature has been removed

/* eslint-disable max-classes-per-file, @typescript-eslint/no-unused-vars */

import { BadRequestException, ForbiddenException } from '@nestjs/common';
import { ApiOperationOptions, ApiResponse, ApiResponseOptions } from '@nestjs/swagger';

import { ApiException } from '../../lib';

jest.mock('@nestjs/swagger', () => {
  const {
    ApiOperation: ActualApiOperation,
    ApiResponse: ActualApiResponse,
    ...otherImplementations
  } = jest.requireActual('@nestjs/swagger');

  return {
    ...otherImplementations,
    ApiResponse: jest.fn((options: ApiResponseOptions) => ActualApiResponse(options)),
    ApiOperation: (options: ApiOperationOptions) => ActualApiOperation(options),
  };
});

describe('Decorator', () => {
  const ApiResponseMock = ApiResponse as jest.Mock<typeof ApiResponse>;

  beforeEach(() => {
    ApiResponseMock.mockClear();
  });

  describe('@ApiException - single exception', () => {
    describe('given valid NestJS built in exception without template or description', () => {
      it('should use the default template', () => {
        class Ignore {
          @ApiException(BadRequestException)
          test() {
            return;
          }
        }

        expect(ApiResponseMock.mock.calls[0][0]).toMatchSnapshot();
      });
    });

    describe('given valid NestJS built in exception (forbidden exception) without template or description', () => {
      it('should use the default template including the error property', () => {
        class Ignore {
          @ApiException(ForbiddenException)
          test() {
            return;
          }
        }

        expect(ApiResponseMock.mock.calls[0][0]).toMatchSnapshot();
      });
    });

    describe('given valid NestJS built in exception without template but with description', () => {
      it('should use the default template', () => {
        class Ignore {
          @ApiException(BadRequestException, { description: 'This is a test' })
          test() {
            return;
          }
        }

        expect(ApiResponseMock.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
});
