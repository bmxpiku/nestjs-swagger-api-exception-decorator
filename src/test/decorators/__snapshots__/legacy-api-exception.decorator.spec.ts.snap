// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decorator @ApiException - single exception given valid NestJS built in exception (forbidden exception) without template or description should use the default template including the error property 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "ForbiddenException": Object {
          "description": "Forbidden",
          "value": Object {
            "error": "Forbidden",
            "message": "Forbidden",
            "statusCode": 403,
          },
        },
      },
      "schema": Object {
        "description": undefined,
        "properties": Object {
          "error": Object {
            "example": "Forbidden",
            "type": "string",
          },
          "message": Object {
            "example": "Forbidden",
            "type": "string",
          },
          "statusCode": Object {
            "example": 403,
            "type": "number",
          },
        },
        "required": Array [
          "statusCode",
          "message",
        ],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 403,
  "type": undefined,
}
`;

exports[`Decorator @ApiException - single exception given valid NestJS built in exception without template but with description should use the default template 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "BadRequestException": Object {
          "description": "This is a test",
          "value": Object {
            "error": "Bad Request",
            "message": "Bad Request",
            "statusCode": 400,
          },
        },
      },
      "schema": Object {
        "description": "This is a test",
        "properties": Object {
          "error": Object {
            "example": "Bad Request",
            "type": "string",
          },
          "message": Object {
            "example": "Bad Request",
            "type": "string",
          },
          "statusCode": Object {
            "example": 400,
            "type": "number",
          },
        },
        "required": Array [
          "statusCode",
          "message",
        ],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 400,
  "type": undefined,
}
`;

exports[`Decorator @ApiException - single exception given valid NestJS built in exception without template or description should use the default template 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "BadRequestException": Object {
          "description": "Bad Request",
          "value": Object {
            "error": "Bad Request",
            "message": "Bad Request",
            "statusCode": 400,
          },
        },
      },
      "schema": Object {
        "description": undefined,
        "properties": Object {
          "error": Object {
            "example": "Bad Request",
            "type": "string",
          },
          "message": Object {
            "example": "Bad Request",
            "type": "string",
          },
          "statusCode": Object {
            "example": 400,
            "type": "number",
          },
        },
        "required": Array [
          "statusCode",
          "message",
        ],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 400,
  "type": undefined,
}
`;

exports[`Decorator @TemplatedApiException given exception array should use the default template 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "BadRequestException": Object {
          "description": "Bad Request",
          "value": Object {},
        },
      },
      "schema": Object {
        "description": undefined,
        "properties": Object {},
        "required": Array [],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 400,
  "type": undefined,
}
`;

exports[`Decorator @TemplatedApiException given single exception should not print a warning anymore because of duplicate class name 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "BadRequestException": Object {
          "description": "Bad Request",
          "value": Object {},
        },
      },
      "schema": Object {
        "description": undefined,
        "properties": Object {},
        "required": Array [],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 400,
  "type": undefined,
}
`;

exports[`Decorator @TemplatedApiException given single exception should use the default template 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "examples": Object {
        "BadRequestException": Object {
          "description": "Bad Request",
          "value": Object {},
        },
      },
      "schema": Object {
        "description": undefined,
        "properties": Object {},
        "required": Array [],
        "type": "object",
      },
    },
  },
  "description": "",
  "isArray": undefined,
  "status": 400,
  "type": undefined,
}
`;
